src/hotspot/share/jfr/leakprofiler/utilities/granularTimer.cpp

```diff
@@ -36,8 +36,8 @@ void GranularTimer::start(jlong duration_ticks, long granularity) {
   _granularity = granularity;
   _counter = granularity;
   _start_time_ticks = JfrTicks::now();
-  const jlong end_time_ticks = _start_time_ticks.value() + duration_ticks;
-  _finish_time_ticks = end_time_ticks < 0 ? JfrTicks(max_jlong) : JfrTicks(end_time_ticks);
+  const julong end_time_ticks = (julong)_start_time_ticks.value() + (julong)duration_ticks;
+  _finish_time_ticks = end_time_ticks > (julong)max_jlong ? JfrTicks(max_jlong) : JfrTicks(end_time_ticks);
   _finished = _finish_time_ticks == _start_time_ticks;
```

Comment:
Isn't there a much better way to fix this?

Reply:
If you change 

`if (JfrTicks::now() > _finish_time_ticks) {`

to

`if (JfrTicks::now() - _finish_time_ticks > 0) {`

you get natural wrap-around semantics. As long as the total interval is less than half the total julong range. And you don't need to worry about `max_jlong`, etc.

Reply:
The problem here is Java code is passing a value of max_jlong for duration_ticks.  I don't see how changing the logic (in is_finished()?) helps with that.